### PR TITLE
Adds Windows PnP Devices and Driver status query APIs

### PR DESCRIPTION
**Pull request by nefarius**
Adds utility functions to query device and driver status' on Windows.

### CODE REVIEW

File: src/windows/pnpdevice.c [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+	*status = MTY_PNP_DEVICE_STATUS_NOT_FOUND;
+
+	SP_DEVINFO_DATA spDevInfoData = {0};
+
+	const HDEVINFO hDevInfo = SetupDiGetClassDevsA(
+		classGuid,
+		NULL,
+		NULL,
+		DIGCF_PRESENT
+	);
```
Comment (Namaneo) [inline src/windows/pnpdevice.c]:
```suggestion
	const HDEVINFO hDevInfo = SetupDiGetClassDevsA(classGuid, NULL, NULL, DIGCF_PRESENT);
```

File: src/windows/pnpdevice.c [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+			if (GetLastError() == ERROR_INVALID_DATA)
+				break;
+			
+			if (GetLastError() == ERROR_INSUFFICIENT_BUFFER) {
+				if (buffer)
+					LocalFree(buffer);
+				buffer = LocalAlloc(LPTR, bufferSize);
+			} else {
+				
+				goto except;
```
Comment (Namaneo) [inline src/windows/pnpdevice.c]:
```suggestion

			} else {
				goto except;
```

File: src/windows/pnpdevice.c [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+		uint32_t instance = 0;
+
+		// get hardware ID(s) property
+		while (!SetupDiGetDeviceRegistryPropertyA(hDevInfo,
+			&spDevInfoData,
+			SPDRP_HARDWAREID,
+			&type,
+			(PBYTE)buffer,
+			bufferSize,
+			&bufferSize)) {
```
Comment (Namaneo) [inline src/windows/pnpdevice.c]:
```suggestion
		while (!SetupDiGetDeviceRegistryPropertyA(hDevInfo, &spDevInfoData, SPDRP_HARDWAREID, &type, (PBYTE) buffer, bufferSize, &bufferSize)) {
```

File: src/windows/pnpdevice.c [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+		}
+	}
+
+	// device not found but no enumeration error
+	if (!found && GetLastError() == ERROR_SUCCESS) {
+		*status = MTY_PNP_DEVICE_STATUS_NOT_FOUND;
+		succeeded = true;
+	}
+
+except:
```
Comment (Namaneo) [inline src/windows/pnpdevice.c]:
```suggestion
	except:

```

File: src/windows/pnpdevice.c [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+	*status = MTY_PNP_DEVICE_STATUS_NOT_FOUND;
+
+	SP_DEVINFO_DATA spDevInfoData = {0};
+
+	const HDEVINFO hDevInfo = SetupDiGetClassDevsA(
+		interfaceGuid,
+		NULL,
+		NULL,
+		DIGCF_PRESENT | DIGCF_DEVICEINTERFACE
+	);
```
Comment (Namaneo) [inline src/windows/pnpdevice.c]:
```suggestion
	const HDEVINFO hDevInfo = SetupDiGetClassDevsA(interfaceGuid, NULL, NULL, DIGCF_PRESENT | DIGCF_DEVICEINTERFACE);
```

File: src/windows/pnpdevice.c [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+	deviceInterfaceData.cbSize = sizeof(deviceInterfaceData);
+
+	// enumerate device instances
+	while (SetupDiEnumDeviceInterfaces(
+		hDevInfo,
+		NULL,
+		interfaceGuid,
+		memberIndex++,
+		&deviceInterfaceData
+	)) {
```
Comment (Namaneo) [inline src/windows/pnpdevice.c]:
```suggestion
	while (SetupDiEnumDeviceInterfaces(hDevInfo, NULL, interfaceGuid, memberIndex++, &deviceInterfaceData)) {
```

File: src/windows/pnpdevice.c [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+
+		// get content
+		if (SetupDiGetDeviceInterfaceDetail(
+			hDevInfo,
+			&deviceInterfaceData,
+			detailDataBuffer,
+			requiredSize,
+			&requiredSize,
+			NULL
+		)) {
```
Comment (Namaneo) [inline src/windows/pnpdevice.c]:
```suggestion
		if (SetupDiGetDeviceInterfaceDetail(hDevInfo, &deviceInterfaceData, detailDataBuffer, requiredSize, &requiredSize, NULL)) {
```

File: src/windows/pnpdevice.c [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+			DEVPROPTYPE type = 0;
+			// get required property size in bytes
+			CONFIGRET ret = CM_Get_Device_Interface_Property(
+				detailDataBuffer->DevicePath,
+				&DEVPKEY_Device_InstanceId,
+				&type,
+				NULL,
+				&bytes,
+				0
+			);
```
Comment (Namaneo) [inline src/windows/pnpdevice.c]:
```suggestion
			CONFIGRET ret = CM_Get_Device_Interface_Property(detailDataBuffer->DevicePath, &DEVPKEY_Device_InstanceId, &type, NULL, &bytes, 0);
```

File: src/windows/pnpdevice.c [hunk 9]
```diff
@@ -0,0 +1,414 @@
...
+
+			// get property content
+			ret = CM_Get_Device_Interface_Property(
+				detailDataBuffer->DevicePath,
+				&DEVPKEY_Device_InstanceId,
+				&type,
+				(PBYTE)instanceId,
+				&bytes,
+				0
+			);
```
Comment (Namaneo) [inline src/windows/pnpdevice.c]:
```suggestion
			ret = CM_Get_Device_Interface_Property(detailDataBuffer->DevicePath, &DEVPKEY_Device_InstanceId, &type, (PBYTE) instanceId, &bytes, 0);
```

File: src/windows/pnpdevice.c [hunk 10]
```diff
@@ -0,0 +1,414 @@
...
+		*status = MTY_PNP_DEVICE_STATUS_NOT_FOUND;
+		succeeded = true;
+		SetLastError(ERROR_SUCCESS);
+	}
+
+except:
+	if (instanceId)
+		free(instanceId);
+	if (detailDataBuffer)
+		free(detailDataBuffer);
```
Comment (Namaneo) [inline src/windows/pnpdevice.c]:
```suggestion
	except:

	free(instanceId);
	free(detailDataBuffer);
```

File: src/windows/pnpdevice.c [hunk 11]
```diff
@@ -0,0 +1,414 @@
...
+	bool found = false, succeeded = false;
+
+	SP_DEVINFO_DATA spDevInfoData = {0};
+
+	const HDEVINFO hDevInfo = SetupDiGetClassDevsA(
+		classGuid,
+		NULL,
+		NULL,
+		DIGCF_PRESENT
+	);
```
Comment (Namaneo) [inline src/windows/pnpdevice.c]:
```suggestion
	const HDEVINFO hDevInfo = SetupDiGetClassDevsA(classGuid, NULL, NULL, DIGCF_PRESENT);
```

File: src/windows/pnpdevice.c [hunk 12]
```diff
@@ -0,0 +1,414 @@
...
+		uint32_t instance = 0;
+
+		// get hardware ID(s) property
+		while (!SetupDiGetDeviceRegistryPropertyA(hDevInfo,
+			&spDevInfoData,
+			SPDRP_HARDWAREID,
+			NULL,
+			(PBYTE)buffer,
+			bufferSize,
+			&bufferSize)) {
```
Comment (Namaneo) [inline src/windows/pnpdevice.c]:
```suggestion
		while (!SetupDiGetDeviceRegistryPropertyA(hDevInfo, &spDevInfoData, SPDRP_HARDWAREID, NULL, (PBYTE) buffer, bufferSize, &bufferSize)) {
```

File: src/windows/pnpdevice.c [hunk 13]
```diff
@@ -0,0 +1,414 @@
...
+			// get driver key name size
+			SetupDiGetDeviceRegistryPropertyA(
+				hDevInfo,
+				&spDevInfoData,
+				SPDRP_DRIVER,
+				NULL,
+				NULL,
+				bufferSize,
+				&bufferSize
+			);
```
Comment (Namaneo) [inline src/windows/pnpdevice.c]:
```suggestion
			SetupDiGetDeviceRegistryPropertyA(hDevInfo, &spDevInfoData, SPDRP_DRIVER, NULL, NULL, bufferSize, &bufferSize);
```

File: src/windows/pnpdevice.c [hunk 14]
```diff
@@ -0,0 +1,414 @@
...
+			// get driver key name content
+			if (!SetupDiGetDeviceRegistryPropertyA(
+				hDevInfo,
+				&spDevInfoData,
+				SPDRP_DRIVER,
+				NULL,
+				(PBYTE)buffer,
+				bufferSize,
+				&bufferSize
+			)) {
```
Comment (Namaneo) [inline src/windows/pnpdevice.c]:
```suggestion
			if (!SetupDiGetDeviceRegistryPropertyA(hDevInfo, &spDevInfoData, SPDRP_DRIVER, NULL, (PBYTE) buffer, bufferSize, &bufferSize)) {
```

File: src/windows/pnpdevice.c [hunk 15]
```diff
@@ -0,0 +1,414 @@
...
+			bufferSize = 0;
+			RegGetValueA(
+				hKey,
+				NULL,
+				DRIVER_VERSION_VALUE_NAME,
+				RRF_RT_REG_SZ | RRF_ZEROONFAILURE,
+				NULL,
+				NULL,
+				&bufferSize
+			);
```
Comment (Namaneo) [inline src/windows/pnpdevice.c]:
```suggestion
			RegGetValueA(hKey, NULL, DRIVER_VERSION_VALUE_NAME, RRF_RT_REG_SZ | RRF_ZEROONFAILURE, NULL, NULL, &bufferSize);
```

File: src/windows/pnpdevice.c [hunk 16]
```diff
@@ -0,0 +1,414 @@
...
+
+			if (RegGetValueA(
+				hKey,
+				NULL,
+				DRIVER_VERSION_VALUE_NAME,
+				RRF_RT_REG_SZ | RRF_ZEROONFAILURE,
+				NULL,
+				buffer,
+				&bufferSize
+			)) {
```
Comment (Namaneo) [inline src/windows/pnpdevice.c]:
```suggestion
			if (RegGetValueA(hKey, NULL, DRIVER_VERSION_VALUE_NAME, RRF_RT_REG_SZ | RRF_ZEROONFAILURE, NULL, buffer, &bufferSize)) {
```

File: src/windows/pnpdevice.h [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+
+#include <windows.h>
+#include <setupapi.h>
+#include <cfgmgr32.h>
+#include <initguid.h>
+#include <devguid.h>
+#include <devpkey.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <stdio.h>
```
Comment (Namaneo) [inline src/windows/pnpdevice.h]: Is there a reason why you're not including all of those directly in `pnpdevice.c`?

**Review by Namaneo — CHANGES_REQUESTED** [review]

File: src/windows/pnpdevice.h (hunk 17 above)
Comment (nefarius) [reply src/windows/pnpdevice.h]: Originally I thought a header is needed for exposed types but I can migrate it now.

**Comment by Kodikuu** [comment]
Please provide stubs to build for other platforms, so that ifdefs can be avoided when these functions are used

File: src/windows/pnpdevice.c [hunk 18]
```diff
@@ -0,0 +1,337 @@
...
+	{
+		DWORD type;
+		LPSTR buffer = NULL;
+		DWORD bufferSize = 0;
+		uint32_t instance = 0;
+
+		// get hardware ID(s) property
+		while (!SetupDiGetDeviceRegistryPropertyA(hDevInfo, &spDevInfoData, SPDRP_HARDWAREID, &type, (PBYTE) buffer, bufferSize, &bufferSize)) {
+			
+			if (GetLastError() == ERROR_INVALID_DATA)
```
Comment (Kodikuu) [inline src/windows/pnpdevice.c]:
```suggestion
		while (!SetupDiGetDeviceRegistryPropertyA(hDevInfo, &spDevInfoData, SPDRP_HARDWAREID, &type, (PBYTE) buffer, bufferSize, &bufferSize)) {
			if (GetLastError() == ERROR_INVALID_DATA)
```

File: src/windows/pnpdevice.c [hunk 19]
```diff
@@ -0,0 +1,337 @@
...
+				break;
+			
+			if (GetLastError() == ERROR_INSUFFICIENT_BUFFER) {
+				if (buffer)
+					LocalFree(buffer);
+				buffer = LocalAlloc(LPTR, bufferSize);
```
Comment (Kodikuu) [inline src/windows/pnpdevice.c]:
```suggestion
					LocalFree(buffer);

				buffer = LocalAlloc(LPTR, bufferSize);
```

File: src/windows/pnpdevice.c [hunk 20]
```diff
@@ -0,0 +1,337 @@
...
+		*status = MTY_PNP_DEVICE_STATUS_NOT_FOUND;
+		succeeded = true;
+	}
+
+	except:
+
+	err = GetLastError();
+	if (hDevInfo)
+		SetupDiDestroyDeviceInfoList(hDevInfo);
+	SetLastError(err);
```
Comment (Kodikuu) [inline src/windows/pnpdevice.c]:
```suggestion
		SetupDiDestroyDeviceInfoList(hDevInfo);

	SetLastError(err);
```

File: src/windows/pnpdevice.c [hunk 21]
```diff
@@ -0,0 +1,337 @@
...
+
+		if (detailDataBuffer) {
+			free(detailDataBuffer);
+			detailDataBuffer = NULL;
+		}
+
+		detailDataBuffer = calloc(requiredSize, 1);
+		if (!detailDataBuffer)
+			goto except;
+		detailDataBuffer->cbSize = sizeof(SP_DEVICE_INTERFACE_DETAIL_DATA);
```
Comment (Kodikuu) [inline src/windows/pnpdevice.c]:
```suggestion
			goto except;

		detailDataBuffer->cbSize = sizeof(SP_DEVICE_INTERFACE_DETAIL_DATA);
```

File: src/windows/pnpdevice.c [hunk 22]
```diff
@@ -0,0 +1,337 @@
...
+
+	except:
+
+	free(instanceId);
+	free(detailDataBuffer);
+
+	err = GetLastError();
+	if (hDevInfo)
+		SetupDiDestroyDeviceInfoList(hDevInfo);
+	SetLastError(err);
```
Comment (Kodikuu) [inline src/windows/pnpdevice.c]:
```suggestion
		SetupDiDestroyDeviceInfoList(hDevInfo);

	SetLastError(err);
```

File: src/windows/pnpdevice.c [hunk 23]
```diff
@@ -0,0 +1,337 @@
...
+
+	for (DWORD i = 0; SetupDiEnumDeviceInfo(hDevInfo, i, &spDevInfoData); i++) {
+		LPSTR buffer = NULL;
+		DWORD bufferSize = 0;
+		uint32_t instance = 0;
+
+		// get hardware ID(s) property
+		while (!SetupDiGetDeviceRegistryPropertyA(hDevInfo, &spDevInfoData, SPDRP_HARDWAREID, NULL, (PBYTE) buffer, bufferSize, &bufferSize)) {
+			
+			if (GetLastError() == ERROR_INVALID_DATA)
```
Comment (Kodikuu) [inline src/windows/pnpdevice.c]:
```suggestion
		while (!SetupDiGetDeviceRegistryPropertyA(hDevInfo, &spDevInfoData, SPDRP_HARDWAREID, NULL, (PBYTE) buffer, bufferSize, &bufferSize)) {
			if (GetLastError() == ERROR_INVALID_DATA)
```

File: src/windows/pnpdevice.c [hunk 24]
```diff
@@ -0,0 +1,337 @@
...
+				break;
+
+			if (GetLastError() == ERROR_INSUFFICIENT_BUFFER) {
+				if (buffer)
+					LocalFree(buffer);
+				buffer = LocalAlloc(LPTR, bufferSize);
```
Comment (Kodikuu) [inline src/windows/pnpdevice.c]:
```suggestion
					LocalFree(buffer);

				buffer = LocalAlloc(LPTR, bufferSize);
```

File: src/windows/pnpdevice.c [hunk 25]
```diff
@@ -0,0 +1,337 @@
...
+			// get driver key name content
+			if (!SetupDiGetDeviceRegistryPropertyA(hDevInfo, &spDevInfoData, SPDRP_DRIVER, NULL, (PBYTE) buffer, bufferSize, &bufferSize)) {
+				LocalFree(buffer);
+				goto except;
+			}
+
+			HKEY hKey;
+			CHAR subKey[MAX_PATH];
+			// build driver key path
+			(void)sprintf_s(subKey, MAX_PATH, "SYSTEM\\CurrentControlSet\\Control\\Class\\%s", buffer);
```
Comment (Kodikuu) [inline src/windows/pnpdevice.c]:
```suggestion
			sprintf_s(subKey, MAX_PATH, "SYSTEM\\CurrentControlSet\\Control\\Class\\%s", buffer);
```

File: src/windows/pnpdevice.c [hunk 26]
```diff
@@ -0,0 +1,337 @@
...
+
+	// device not found but no enumeration error
+	if (!found && GetLastError() == ERROR_SUCCESS)
+		succeeded = false;
+
+except:
+	err = GetLastError();
+	if (hDevInfo)
+		SetupDiDestroyDeviceInfoList(hDevInfo);
+	SetLastError(err);
```
Comment (Kodikuu) [inline src/windows/pnpdevice.c]:
```suggestion
		SetupDiDestroyDeviceInfoList(hDevInfo);

	SetLastError(err);
```

**Review by Kodikuu — CHANGES_REQUESTED** [review]

**Comment by Namaneo** [comment]
Closing as it was moved to Parsec internals

**Pull request closed without merging**